### PR TITLE
Remove unnecessary templates

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The function `arrayAUC` was using heavy C++ templates.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 2f6e66c8e3809d80bcab0afe7bf5d9d03761640d with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53183/2f6e66c8e3809d80bcab0afe7bf5d9d03761640d/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/array/arrayAUC.cpp [hunk 1]
```diff
@@ -129,18 +121,84 @@ class ArrayAUCImpl
...
+
+public:
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 2; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo &) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        // Basic type check
+        std::vector<DataTypePtr> nested_types(2, nullptr);
```
Comment (antaljanosbenjamin) [inline src/Functions/array/arrayAUC.cpp:155]: I don't see where these nested types are used apart from collecting them.

File: src/Functions/array/arrayAUC.cpp [hunk 2]
```diff
@@ -70,44 +74,32 @@ namespace ErrorCodes
...
-        const T * scores,
-        const U * labels,
-        size_t size)
+    static constexpr auto name = "arrayAUC";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionArrayAUC>(); }
+
+private:
+    static Float64 apply(
+        const IColumn & data1,
+        const IColumn & data2,
```
Comment (antaljanosbenjamin) [inline src/Functions/array/arrayAUC.cpp:86]:
```suggestion
        const IColumn & scores,
        const IColumn & labels,
```

File: src/Functions/array/arrayAUC.cpp [hunk 3]
```diff
@@ -70,44 +74,32 @@ namespace ErrorCodes
...
 
+        size_t size = next_offset - current_offset;
         PODArrayWithStackMemory<ScoreLabel, 1024> sorted_labels(size);
 
         for (size_t i = 0; i < size; ++i)
         {
-            bool label = labels[i] > 0;
-            sorted_labels[i].score = scores[i];
+            bool label = data2.getFloat64(current_offset + i) > 0;
+            sorted_labels[i].score = data1.getFloat64(current_offset + i);
```
Comment (antaljanosbenjamin) [inline src/Functions/array/arrayAUC.cpp:102]: I am not sure what are to consequences of dealing with all types by converting them to `Float64`. Can be there any drawbacks, e.g.: some numeric/enum types cannot be converted to `Float64` or some performance related issues?

File: src/Functions/array/arrayAUC.cpp [hunk 4]
```diff
@@ -129,18 +121,84 @@ class ArrayAUCImpl
...
+            return std::numeric_limits<Float64>::quiet_NaN();
 
-        return static_cast<ResultType>(area) / count_positive / (size - count_positive);
+        return static_cast<Float64>(area) / count_positive / (size - count_positive);
     }
-};
 
+    static void vector(
+        const IColumn & data1,
+        const IColumn & data2,
```
Comment (antaljanosbenjamin) [inline src/Functions/array/arrayAUC.cpp:131]:
```suggestion
        const IColumn & scores,
        const IColumn & labels,
```

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]
The test failures are kind of deterministic, but I think they are not connected to your changes, but I am not 100% sure.

File: src/Functions/array/arrayAUC.cpp (hunk 3 above)
Comment (alexey-milovidov) [reply src/Functions/array/arrayAUC.cpp]: About performance - this is a rare, obscure function, and its inner loop is already complex enough, so I don't expect these virtual calls will make a difference.

**Comment by alexey-milovidov** [comment]
@antaljanosbenjamin, it looks like it is ready, but we need to fix the Performance test and Upgrade check.

File: src/Functions/array/arrayAUC.cpp (hunk 4 above)
Comment (antaljanosbenjamin) [reply src/Functions/array/arrayAUC.cpp:131]: You only changed `data1` and `data2` in one of the functions.

**Comment by antaljanosbenjamin** [comment]
https://github.com/ClickHouse/ClickHouse/pull/53183#discussion_r1288159101

You haven't addressed this yet I think.

**Review by antaljanosbenjamin — APPROVED** [review]

**Pull request merged**